### PR TITLE
GEOMESA-2263 Accurate dwithin filters

### PR DESCRIPTION
**Pull request by elahrvivaz**
* New FastDWithin filter that compares using meters
* DWithin filters are not re-written to degrees
* Requires distributed-runtime jar to be updated

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.locationtech.geomesa.utils.geotools.GeometryUtils
+import org.opengis.filter.FilterVisitor
+import org.opengis.filter.MultiValuedFilter.MatchAction
+import org.opengis.filter.expression.{Expression, Literal}
+import org.opengis.filter.spatial.DWithin
+
+object FastDWithin {
+
+  private val calculators = new ThreadLocal[GeodeticCalculator] {
+    override def initialValue: GeodeticCalculator = new GeodeticCalculator()
```
Comment (jnh5y) [inline geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala]: I think this duplicates the code/ideas of GeometricDistanceFunctions.fastDistance.

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala [hunk 2]
```diff
@@ -51,106 +51,59 @@ object FilterHelper {
...
     val geom = prop.literal.evaluate(null, classOf[Geometry])
     if (geom.getUserData == SafeGeomString) {
       op // we've already visited this geom once
     } else {
       // check for null or empty attribute and replace with default geometry name
-      val attribute = Option(prop.name).filterNot(_.isEmpty).getOrElse(if (sft == null) null else sft.getGeomField)
+      val attribute = Option(prop.name).filterNot(_.isEmpty).orElse(Option(sft).map(_.getGeomField)).orNull
       // copy the geometry so we don't modify the original
-      val geomCopy = gf.createGeometry(geom)
+      val geomCopy = geom.getFactory.createGeometry(geom)
```
Comment (jnh5y) [inline geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala]: hehehe.... Does Geometry have a clone method for this kind of use?

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala]: is that a geotools thing?

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala]: hmm, it actually does have a clone() implementation - not sure why we didn't use that to begin with. this change was mainly just to preserve any precision models associated with the geom, probably irrelevant...

File: geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala]: hmm, that doesn't seem to be fast haha, it just uses the geodetic calculator

Comment (jnh5y) [reply geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/expression/FastDWithin.scala]: https://github.com/locationtech/geomesa/blob/master/geomesa-spark/geomesa-spark-sql/src/main/scala/org/locationtech/geomesa/spark/GeometricDistanceFunctions.scala

**Pull request merged**